File: STRIKE-GOLDD/StrikePy/functions/elim_and_recalc.py
"""
    Function to check the observability / identifiability of each variable
"""

def elim_and_recalc(unmeas_xred_indices, rangoinicial, numonx, p, x, unidflag, w1vector, *args):
    ################################################################################
    # Imports:
    import options
    import sympy as sym
    from sympy import Matrix
    import symbtools as st
    from functions.rationalize import rationalize_all_numbers
    ################################################################################
    numonx = rationalize_all_numbers(Matrix(numonx))
    # Depending on the number of arguments you pass to the function, there are two cases:

    # called when there is no 'w'
    if len(args) == 0:
        pred = p
        xred = x
        wred = w1vector
        identifiables = []
        obs_states = []
        obs_inputs = []
        q = len(pred)
        n = len(xred)
        nw = len(wred)

    # called when there are 'w'
    if len(args) == 3:
        pred = p
        xred = x
        wred = w1vector
        identifiables = args[0]
        obs_states = args[1]
        obs_inputs = args[2]
        q = len(pred)
        n = len(xred)
        nw = len(wred)

    r = sym.shape(Matrix(numonx))[1] # before: q+n+nw; but with unknown inputs there may also be derivatives
    new_ident_pars = identifiables
    new_nonid_pars = []
    new_obs_states = obs_states
    new_unobs_states = []
    new_obs_in = obs_inputs
    new_unobs_in = []

    #========================================================================
    # ELIMINATE A PARAMETER:
    #========================================================================
    # At each iteration we remove a different column (= parameter) from onx:
    for ind in range(q): # for each parameter of p...
        if q <= 1: # check if the parameter has already been marked as identifiable
            isidentifiable = pred[ind] in identifiables
        else:
            isidentifiable = any(pred[ind] in arr for arr in identifiables)
        if isidentifiable:
            print('\n Parameter {} has already been classified as identifiable.'.format(pred[ind]))
        else:
            indices = []
            for i in range(r):
                indices.append(i)
            indices.pop(n + ind)
            column_del_numonx = Matrix(numonx).col(indices) # one column is removed
            num_rank = st.generic_rank(Matrix(column_del_numonx)) # the range is calculated without that column
            if num_rank == rangoinicial:
                if unidflag == 1:
                    print('\n    => Parameter {} is structurally unidentifiable'.format(pred[ind]))
                    new_nonid_pars.append(pred[ind])
                else:
                    print('\n    => We cannot decide about parameter {} at the moment'.format(pred[ind]))
            else:
                print('\n    => Parameter {} is structurally identifiable'.format(pred[ind]))
                new_ident_pars.append(pred[ind])

    # ========================================================================
    # ELIMINATE A STATE:
    # ========================================================================
    # At each iteration we try removing a different state from 'xred':
    if options.checkObser == 1:
        for ind in range(len(unmeas_xred_indices)): # for each unmeasured state
            original_index = unmeas_xred_indices[ind]
            if len(obs_states) <= 1:
                isobservable = xred[original_index] in obs_states
            else:
                isobservable = any(xred[original_index] in arr for arr in obs_states)
            if isobservable:
                print('\n State %s has already been classified as observable.'.format(xred[ind]))
            else:
                indices = []
                for i in range(r):
                    indices.append(i)
                indices.pop(original_index) # remove the column that we want to check
                column_del_numonx = Matrix(numonx).col(indices)
                num_rank = st.generic_rank(Matrix(column_del_numonx))
                if num_rank == rangoinicial:
                    if unidflag == 1:
                        print('\n    => State {} is unobservable'.format(xred[original_index]))
                        new_unobs_states.append(xred[original_index])
                    else: # if this function was called because the necessary number of derivatives was not calculated...
                        print('\n    => We cannot decide about state {} at the moment'.format(xred[original_index]))
                else:
                    print('\n    => State {} is observable'.format(xred[original_index]))
                    new_obs_states.append(xred[original_index])

    # ========================================================================
    # ELIMINATE AN UNKNOWN INPUT:
    # ========================================================================
    # At each iteration we try removing a different column from onx:
    for ind in range(nw): # for each unknown input...
        if len(obs_inputs) <= 1:  # check if the unknown input has already been marked as observable
            isobservable = wred[ind] in obs_inputs
        else:
            isobservable = any(wred[ind] in arr for arr in obs_inputs)
        if isobservable:
            print('\n Input %s has already been classified as observable.'.format(wred[ind]))
        else:
            indices = []
            for i in range(r):
                indices.append(i)
            indices.pop(n+q+ind) # remove the column that we want to check
            column_del_numonx = Matrix(numonx).col(indices)
            num_rank = st.generic_rank(Matrix(column_del_numonx))
            if num_rank == rangoinicial:
                if unidflag == 1:
                    print('\n    => Input {} is unobservable'.format(wred[ind]))
                    new_unobs_in.append(wred[ind])
                else:
                    print('\n    => We cannot decide about input {} at the moment'.format(wred[ind]))
            else:
                print('\n    => Input {} is observable'.format(wred[ind]))
                new_obs_in.append(wred[ind])
    return new_ident_pars, new_nonid_pars, new_obs_states, new_unobs_states, new_obs_in, new_unobs_in

File: STRIKE-GOLDD/functions/elim_and_recalc.m
%--------------------------------------------------------------------------
% Function that determines identifiability of individual parameters one 
% by one, by successive elimination of its column in the identifiability 
% matrix and recalculation of its rank
%--------------------------------------------------------------------------
function [new_ident_pars,new_nonid_pars,new_obs_states,new_unobs_states,new_obs_in,new_unobs_in] = elim_and_recalc(unmeas_xred_indices,rangoinicial,numonx,opts,varargin)

global p x unidflag wlvector

pred          = p;
xred          = x; 
wred          = wlvector;
q             = numel(pred);
n             = numel(xred);
nw            = numel(wred);
qreal         = q;

switch nargin 
    case 4 
        % called when there is no 'w'     
        identifiables = [];
        obs_states    = [];
        obs_inputs    = [];
    case 7 
        % called when there is 'w'
        identifiables = varargin{1};
        obs_states    = varargin{2};    
        obs_inputs    = varargin{3};
end

r  = size(numonx,2); % before: q+n+nw; but with unknown inputs there may also be derivatives
new_ident_pars   = identifiables;
if size(new_ident_pars,2)>size(new_ident_pars,1)
    new_ident_pars = transpose(new_ident_pars);
end
new_nonid_pars   = [];
new_obs_states   = obs_states;
if size(new_obs_states,2)>size(new_obs_states,1)
    new_obs_states = transpose(new_obs_states);
end
new_unobs_states = [];
new_obs_in       = obs_inputs;
if size(new_obs_in,2)>size(new_obs_in,1)
    new_obs_in = transpose(new_obs_in);
end
new_unobs_in     = [];
    

%==========================================================================
% ELIMINATE A PARAMETER:
%==========================================================================
% At each iteration we try removing a different column from onx:
for ind=1:qreal % only the first 'qreal' elements of pred are parameters; the following 'q-qreal' are states considered as parameters 
    isidentifiable = ismember(pred(ind),identifiables);
    if isidentifiable
        fprintf('\n Parameter %s has already been classified as identifiable.',char(pred(ind)))
    else 
        indices = 1:r;
        indices(n+ind) = []; % remove the column whose identifiability we want to check
        num_rank = rank(numonx(:,indices));
        if num_rank == rangoinicial
            if unidflag == 1
            	fprintf('\n    => Parameter %s is structurally unidentifiable',char(pred(ind)));
            	new_nonid_pars = [new_nonid_pars; pred(ind)];
            else
            	fprintf('\n    => We cannot decide about parameter %s at the moment',char(pred(ind)));
            end  
        else
            fprintf('\n    => Parameter %s is structurally identifiable',char(pred(ind)));
            new_ident_pars = [new_ident_pars; pred(ind)];       
        end
    end
end

%==========================================================================
% ELIMINATE A STATE:
%==========================================================================
% At each iteration we try removing a different state from 'xred':
for ind=1:numel(unmeas_xred_indices) % for each unmeasured state
    original_index = unmeas_xred_indices(ind); % in this script, 'original_index' refers to xred
    isobservable = ismember(xred(original_index),obs_states);
    if isobservable
        fprintf('\n State %s has already been classified as observable.',char(xred(original_index)))
    else              
        indices = 1:r;
        indices(original_index) = []; %indices(original_index==unmeas_xred_indices) = [];
        num_rank = rank(numonx(:,indices));
        if num_rank == rangoinicial
            if unidflag == 1
                fprintf('\n    => State %s is unobservable',char(xred(original_index)));
                new_unobs_states = [new_unobs_states; xred(original_index)];
            else
                fprintf('\n    => We cannot decide about state %s at the moment',char(xred(original_index)));
            end 
        else
            fprintf('\n    => State %s is observable',char(xred(original_index)));
            new_obs_states = [new_obs_states; xred(original_index)];  
        end
    end
end


%==========================================================================
% ELIMINATE AN UNKNOWN INPUT:
%==========================================================================
% At each iteration we try removing a different column from onx:
for ind=1:nw 
    isobservable = ismember(wred(ind),obs_inputs);
    if isobservable
        fprintf('\n Input %s has already been classified as observable.',char(wred(ind)))
    else 
        indices = 1:r;
        indices(n+q+ind) = []; % remove the column whose identifiability we want to check
        num_rank = rank(numonx(:,indices));
        if num_rank == rangoinicial
            if unidflag == 1
                fprintf('\n    => Input %s is unobservable',char(wred(ind)));
                new_unobs_in = [new_unobs_in; wred(ind)];
            else
                fprintf('\n    => We cannot decide about input %s at the moment',char(wred(ind)));
            end  
        else
            fprintf('\n    => Input %s is observable',char(wred(ind)));
            new_obs_in = [new_obs_in; wred(ind)];     
        end
    end
end

end
